std/algorithm/sorting.d

```diff
@@ -1423,44 +1423,22 @@ unittest //Issue 16413 - @system comparison function
 
 private size_t getPivot(alias less, Range)(Range r)
 {
-    import std.algorithm.mutation : swapAt;
-
-    // This algorithm sorts the first, middle and last elements of r,
-    // then returns the index of the middle element.  In effect, it uses the
-    // median-of-three heuristic.
-
-    alias pred = binaryFun!(less);
-    immutable len = r.length;
-    immutable size_t mid = len / 2;
-    immutable uint result = ((cast(uint) (pred(r[0], r[mid]))) << 2) |
-                            ((cast(uint) (pred(r[0], r[len - 1]))) << 1) |
-                            (cast(uint) (pred(r[mid], r[len - 1])));
-
-    switch (result)
+    auto mid = r.length / 2;
+    if (r.length < 512)
     {
-        case 0b001:
-            r.swapAt(0, len - 1);
-            r.swapAt(0, mid);
-            break;
-        case 0b110:
-            r.swapAt(mid, len - 1);
-            break;
-        case 0b011:
-            r.swapAt(0, mid);
-            break;
-        case 0b100:
-            r.swapAt(mid, len - 1);
-            r.swapAt(0, mid);
-            break;
-        case 0b000:
-            r.swapAt(0, len - 1);
-            break;
-        case 0b111:
-            break;
-        default:
-            assert(0);
+        if (r.length >= 32)
+            medianOf!less(r, size_t(0), mid, r.length - 1);
+        return mid;
     }
 
+    // The plan here is to take the median of five by taking five elements in
+    // the array, segregate around their median, and return the position of the
+    // third. We choose five equidistant elements in the middle of each
+    // fifthile.
```

Comment:
0 and the end are very seldom in the middle of a fifthile, especially if n >= 512 as required.